### ISSUE TITLE
conditional statements not working with dynamic lists generated using for loop

### ISSUE THREAD

**Issue by vermdeep** [hashicorp/terraform#21996]
### Terraform Version

```
Terraform v0.12.3
+ provider.aws v2.18.0
+ provider.local v1.3.0
```

### Terraform Configuration Files

```hcl
variable "config" {
  type = object({
    container_secrets = list(string)
    custom_container_secrets = list(object({ # ECS secrets map
      name      = string
      valueFrom = string
    }))
  })
}

variable "environment" {}

locals {
  secrets = [
    # Generate ECS secrets map
    for secret in var.config["container_secrets"] :
    {
      name      = secret
      valueFrom = join("/", ["/services", local.service_name, var.environment, secret])
    }
  ]
  service_name = "xyz"
}

resource "local_file" "foo" {
  content  = jsonencode(var.config["custom_container_secrets"] == null ? local.secrets : concat(local.secrets, var.config["custom_container_secrets"]))
  filename = "${path.module}/foo.bar"
}
```

### Expected Behavior
`foo.bar` file to be created with content `[]`

### Actual Behavior
Empty `foo.bar` file  is created.

### Additional Context
On moving the conditional statement to a local variable and then `concat` two list works.
Following is working:
```
variable "config" {
  type = object({
    container_secrets = list(string)
    custom_container_secrets = list(object({ # ECS secrets map
      name      = string
      valueFrom = string
    }))
  })
}

variable "environment" {}

locals {
  secrets = [
    # Generate ECS secrets map
    for secret in var.config["container_secrets"] :
    {
      name      = secret
      valueFrom = join("/", ["/services", local.service_name, var.environment, secret])
    }
  ]
  custom_secrets = var.config["custom_container_secrets"] == null ? [] : var.config["custom_container_secrets"]
  service_name   = "xyz"
}

resource "local_file" "foo" {
  content  = jsonencode(concat(local.secrets, local.custom_secrets))
  filename = "${path.module}/foo.bar"
}

```

Also, the issue is happening in this case only. If `custom_container_secrets` is not null then both the code are working. Also, both the codes are working when `custom_container_secrets` is declared as a variable (not as a key of an object).

**Comment by jbardin** [comment]
Hi @vermdeep,

This looks similar to, and should also be fixed by #21957.
It is turning up in a slightly different place however, so I'll leave this open until we can confirm.

**Comment by jbardin** [comment]
This appears to be working as expected in 0.12.4.
